### PR TITLE
Version 1.76.0

### PR DESCRIPTION
**Pull request by WilliamDEdwards**
# Changes

Implement domain routers.

# Checks

- [x] The version constant is updated in `Client.php`
- [x] The changelog is updated (when applicable)
- [x] The supported Cluster API version is updated in the README (when applicable)

### CODE REVIEW

File: src/Models/DomainRouter.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+class DomainRouter extends ClusterModel implements Model
+{
+    private string $domain;
+    private bool $forceSsl = true;
+    private ?int $virtualHostId = null;
+    private ?int $urlRedirectId = null;
+    private ?int $nodeId = null;
+    private ?int $id = null;
+    private ?int $clusterId = null;
```
Comment (dvdheiden) [inline src/Models/DomainRouter.php:16]: When I look at the documentation, I'm seeing a GET and PUT for this. The PUT requires the `id` and `cluster_id`. A GET always returns those values. So can we remove the nullable property from these properties?

```suggestion
    private int $id;
    private int $clusterId;
```

File: src/Models/DomainRouter.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    public function setNodeId(?int $nodeId): DomainRouter
+    {
+        $this->nodeId = $nodeId;
+
+        return $this;
+    }
+
+    public function getId(): ?int
```
Comment (dvdheiden) [inline src/Models/DomainRouter.php:80]:
```suggestion
    public function getId(): int
```

File: src/Models/DomainRouter.php [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        return $this->id;
+    }
+
+    public function setId(?int $id): DomainRouter
```
Comment (dvdheiden) [inline src/Models/DomainRouter.php:85]:
```suggestion
    public function setId(int $id): DomainRouter
```

File: src/Models/DomainRouter.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        $this->id = $id;
+
+        return $this;
+    }
+
+    public function getClusterId(): ?int
```
Comment (dvdheiden) [inline src/Models/DomainRouter.php:92]:
```suggestion
    public function getClusterId(): int
```

File: src/Models/DomainRouter.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        return $this->clusterId;
+    }
+
+    public function setClusterId(?int $clusterId): DomainRouter
```
Comment (dvdheiden) [inline src/Models/DomainRouter.php:97]:
```suggestion
    public function setClusterId(int $clusterId): DomainRouter
```

**Review by dvdheiden — CHANGES_REQUESTED** [review]

**Comment by dvdheiden** [comment]
@WilliamDEdwards would it be possible to add some documentation about the domain routers? For example: when are they created? etc.

File: src/Models/DomainRouter.php (hunk 1 above)
Comment (WilliamDEdwards) [reply src/Models/DomainRouter.php]: Done in 33fa6ea2939fe757da73a24d061282bf0a67e4f1

File: src/Models/DomainRouter.php (hunk 4 above)
Comment (WilliamDEdwards) [reply src/Models/DomainRouter.php]: Done in 33fa6ea2939fe757da73a24d061282bf0a67e4f1

File: src/Models/DomainRouter.php (hunk 5 above)
Comment (WilliamDEdwards) [reply src/Models/DomainRouter.php]: Done in 33fa6ea2939fe757da73a24d061282bf0a67e4f1

File: src/Models/DomainRouter.php (hunk 2 above)
Comment (WilliamDEdwards) [reply src/Models/DomainRouter.php]: Done in 33fa6ea2939fe757da73a24d061282bf0a67e4f1

File: src/Models/DomainRouter.php (hunk 3 above)
Comment (WilliamDEdwards) [reply src/Models/DomainRouter.php]: Done in 33fa6ea2939fe757da73a24d061282bf0a67e4f1

**Comment by WilliamDEdwards** [comment]
> @WilliamDEdwards would it be possible to add some documentation about the domain routers? For example: when are they created? etc.

We are working on proper documentation for the Cluster API. The current ReDoc document will not be updated.

I wil describe this functionality here.

A request takes an internal route:

- The request traverses the internet to reach the network.
- The request arrives at the load balancers.
- The load balancers proxy the request to the cluster. Possibly to one specific node.
- The request is handled by server software (such as nginx) on the cluster node. 

The 'balancer backend name' attribute determined whether the request should be routed to either:

- Any node in the cluster.
- One specific node.

This attribute belonged to the virtual host and URL redirect models.

This logic fails in the following scenario. I will use `example.com` as an example domain.

- A virtual host exists for `example.com`.
- Requests for `example.com` should not be handled by the virtual host.
- Requests should be handled by an intermediate software layer. For example: Varnish.

In this case, the request should not be routed to the virtual host, but to Varnish. As the 'balancer backend name' belonged to the virtual host and URL redirect models, this was not possible with the previous logic.

If we ever implement Varnish, a 'Varnish VCL' model and a `varnish_vcl_id` attribute (in addition to `virtual_host_id` and `url_redirect_id`) or similar would be added to the domain router model.

The technical implementation of domain routers is as follows:

- Domain routers are created when virtual hosts or URL redirects are created.
- Domain routers are created for the domain and for each server alias.
- When server aliases are removed from virtual hosts or URL redirects, their domain routers are deleted.
- When virtual hosts or URL redirects are deleted, their domain routers are deleted.
- To route incoming requests to a specific node, set the node ID attribute.
- 1 domain router exists for each domain.
- Either virtual host ID or URL redirect ID is set.

**Comment by dvdheiden** [comment]
Is it OK for you when I put this information into the wiki of this GitHub repo for now?

**Comment by WilliamDEdwards** [comment]
> Is it OK for you when I put this information into the wiki of this GitHub repo for now?

Yes, that is OK.

**Review by dvdheiden — APPROVED** [review]

**Pull request merged**